File: src/cmcandy/C_Language_Answers/_0086.c
# include <stdio.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *partition(struct ListNode *head, int x)
{
    struct ListNode *small = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode *big = (struct ListNode *)malloc(sizeof(struct ListNode));

    //定义仨指针
    struct ListNode *pSmall = small;
    struct ListNode *pBig = big;

    while (head)
    {
        if (head->val < x)
        {
            pSmall->next = head;
            pSmall = pSmall->next;
        }
        else
        {
            pBig->next = head;
            pBig = pBig->next;
        }
        head = head->next;
    }

    pBig->next = NULL;
    pSmall->next = big->next;

    return small->next;
}

File: src/cmcandy/java_language_answers/_0086.java
package cmcandy.java_language_answers;

public class _0086 {
    /**
     * Definition for singly-linked list. public class ListNode { int val; ListNode
     * next; ListNode(int x) { val = x; } }
     */
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    class Solution {
        public ListNode partition(ListNode head, int x) {
            // xiao da
            ListNode small = new ListNode(-1);
            ListNode big = new ListNode(-1);

            ListNode pSmall = small;
            ListNode pBig = big;

            while (head != null) {
                if (head.val < x) {
                    pSmall.next = head;
                    pSmall = pSmall.next;
                } else {
                    pBig.next = head;
                    pBig = pBig.next;
                }
                head = head.next;
            }
            pBig.next = null;
            pSmall.next = big.next;
            return small.next;
        }
    }
}